### ISSUE TITLE
Broken XML_ID uniqueness validation for tests

### ISSUE THREAD

**Issue by Mathis-Z** [openHPI/codeharbor#1247]
There is a validation for tests to make sure there cannot be two tests with the same XML_ID for the same task. However this validation appears to be broken because I can just create tasks with tests that have the same XML_ID.
I guess Rails performs the validation of both tests before they are saved to the database and each validation check does not find the other test in the database so they just pass and get written to database successfully.
We should check if there are similar issues in other models.

EDIT: You can create a task like that but you cannot edit it afterwards because then the tests are in the database and the validation fails.

**Comment by MrSerth** [comment]
Indeed, I can confirm that this is a bug in the current implementation. We also noticed that when implementing contributions, but just added corresponding comments in the `fs_task_contrib` branch, yet without a ticket nor fixing it. From the tests I've made, I think this is an issue for several models, not just `Test`.

**Comment by kkoehn** [comment]
It seems to be a bug in Rails. The corresponding issue (20676) was opened in 2015. There are multiple workarounds:
- our existing `unique_xml_id` validation for `TaskFiles` seems to work fine (since it works with the objects)
- monkeypatch the uniqueness validator
- create a (kind of) generic validator for that problem (https://github.com/rails/rails/issues/20676#issuecomment-1068020772)
- create the correct database constraints (!) and use a gem that converts a database constraint exceptions to a validation message (https://github.com/toptal/database_validations)

As far as I understand the issue, it only affects objects that are saved through nested_attributes. Since we are only using nested_attributes for `ModelSolutions`, `Tests` and `TaskFiles` and the `TaskFiles` need a specific validator anyway, we only need to find a solution for `Tests` and `ModelSolutions`.

I suggest we add the necessary database constraints and a validator that we can use for both models.

**Comment by Mathis-Z** [comment]
Yes, that sounds good.

**Comment by kkoehn** [comment]
I'm not able to reproduce this bug currently. It may have gotten fixed by https://github.com/openHPI/codeharbor/pull/1103 and the added `ParentValidation`.

**Comment by MrSerth** [comment]
I just tested on the latest `main` branch (commit 5a9d6d3b) and was able to reproduce the issue. Below is a reproduction video.

https://github.com/user-attachments/assets/6ec70c67-41b7-4ecb-bd4e-87e171ffddb9

**Comment by kkoehn** [comment]
Ah I misunderstood and created `TaskFiles` instead of `Tests` with the same id. Thanks!
